tests/integration_tests/performance/test_huge_pages.py

```diff
@@ -114,6 +118,73 @@ def test_hugetlbfs_snapshot(
     check_hugetlbfs_in_use(vm.firecracker_pid, "/anon_hugepage")
 
 
+@pytest.mark.skipif(
+    global_props.host_linux_version == "4.14",
+    reason="MFD_HUGETLB | MFD_ALLOW_SEALING only supported on kernels >= 4.16",
+)
+@pytest.mark.parametrize("huge_pages", HugePagesConfig)
+def test_ept_violation_count(
+    microvm_factory,
+    guest_kernel_linux_5_10,
+    rootfs_ubuntu_22,
+    uffd_handler_paths,
+    metrics,
+    huge_pages,
+):
+    """
+    Tests hugetlbfs snapshot restore with a UFFD handler that pre-faults the entire guest memory
+    on the first page fault. Records metrics about the number of EPT_VIOLATIONS encountered by KVM.
+    """
+
+    ### Create Snapshot ###
+    vm = microvm_factory.build(guest_kernel_linux_5_10, rootfs_ubuntu_22)
+    vm.memory_monitor = None
+    vm.spawn()
+    vm.basic_config(huge_pages=huge_pages, mem_size_mib=256)
+    vm.add_net_iface()
+    vm.start()
+
+    metrics.set_dimensions(
+        {
+            "performance_test": "test_hugetlbfs_snapshot",
+            "huge_pages_config": str(huge_pages),
+            **vm.dimensions,
+        }
+    )
+
+    # Wait for microvm to boot. Also touch a bunch of pages to grow the guest's page tables.
+    rc, _, _ = vm.ssh.run("head -c 128M /dev/zero | tail")
+    assert not rc
+
+    snapshot = vm.snapshot_full()
+
+    vm.kill()
+
+    ### Restore Snapshot ###
+    vm = microvm_factory.build()
+    vm.jailer.daemonize = False
+    vm.jailer.extra_args.update({"no-seccomp": None})
+    vm.spawn()
+
+    # Spawn page fault handler process.
+    _pf_handler = spawn_pf_handler(
+        vm, uffd_handler_paths["fault_all_handler"], snapshot.mem
+    )
+
+    with ftrace_events("kvm:*"):
+        vm.restore_from_snapshot(snapshot, resume=True, uffd_path=SOCKET_PATH)
+
+        # Verify if guest can run commands, and also touch a bunch of pages to trigger EPT_VIOLATION exits
+        rc, _, _ = vm.ssh.run("head -c 128M /dev/zero | tail")
```

Comment:
do we need to do this both before and after taking a snapshot?

Reply:
We need to do it before to "grow" the snapshot (e.g. make sure the guest's page tables have a lot of entries that can potentially trigger EPT_VIOLATION exits after restore). We need to do it again after restore to actually trigger ept violations.

Although thinking about it now, this should probably keep around the same memory area (e.g. allocate and touch pages before taking snapshot, keep allocation alive across restore point, touch same pages again after restore). ugh.

Reply:
ok, fair enough, thx.